lib/pacer.go

```diff
@@ -0,0 +1,216 @@
+		return 0, false
+	}
+	interval := uint64(cp.Per.Nanoseconds() / int64(cp.Freq))
+	if math.MaxInt64/interval < hits {
+		// We would overflow delta if we continued, so stop the attack.
+		return 0, true
+	}
+	delta := time.Duration((hits + 1) * interval)
+	// Zero or negative durations cause time.Sleep to return immediately.
+	return delta - elapsed, false
+}
+
+// hitsPerNs returns the attack rate this ConstantPacer represents, in
+// fractional hits per nanosecond.
+func (cp ConstantPacer) hitsPerNs() float64 {
+	return float64(cp.Freq) / float64(cp.Per)
+}
+
+const (
+	// MeanUp is a SinePacer Offset that causes the attack to start
+	// at the Mean attack rate and increase towards the peak.
+	MeanUp float64 = 0
+	// Peak is a SinePacer Offset that causes the attack to start
+	// at the peak (maximum) attack rate and decrease towards the Mean.
+	Peak = math.Pi / 2
+	// MeanDown is a SinePacer Offset that causes the attack to start
+	// at the Mean attack rate and decrease towards the trough.
+	MeanDown = math.Pi
+	// Trough is a SinePacer Offset that causes the attack to start
+	// at the trough (minimum) attack rate and increase towards the Mean.
+	Trough = 3 * math.Pi / 2
+)
+
+// SinePacer is a Pacer that describes attack request rates with the equation:
+//   R = MA sin(O+(2𝛑/P)t)
+// Where:
+//   R = Instantaneous attack rate at elapsed time t, hits per nanosecond
+//   M = Mean attack rate over period P, sp.Mean, hits per nanosecond
+//   A = Amplitude of sine wave, sp.Amp, hits per nanosecond
+//   O = Offset of sine wave, sp.StartAt, radians
+//   P = Period of sine wave, sp.Period, nanoseconds
+//   t = Elapsed time since attack start, nanoseconds
+//
+// Many thanks to http://ascii.co.uk/art/sine and "sps" for the ascii here :-)
+//
+//  Mean -|         ,-'''-.
+//  +Amp  |      ,-'   |   `-.
+//        |    ,'      |      `.       O=𝛑
+//        |  ,'      O=𝛑/2      `.     MeanDown
+//        | /        Peak         \   /
+//        |/                       \ /
+//  Mean -+-------------------------\--------------------------> t
+//        |\                         \                       /
+//        | \                         \       O=3𝛑/2        /
+//        |  O=0                       `.     Trough      ,'
+//        |  MeanUp                      `.      |      ,'
+//  Mean  |                                `-.   |   ,-'
+//  -Amp -|                                   `-,,,-'
+//        |<-------------------- Period --------------------->|
+//
+// This equation is integrated with respect to time to derive the expected
+// number of hits served at time t after the attack began:
+//   H = Mt - (AP/2𝛑)cos(O+(2𝛑/P)t) + (AP/2𝛑)cos(O)
+// Where:
+//   H = Total number of hits triggered during t
+type SinePacer struct {
+	// The period of the sine wave, e.g. 20*time.Minute
+	// MUST BE > 0
+	Period time.Duration
+	// The mid-point of the sine wave in freq-per-Duration,
+	// MUST BE > 0
+	Mean ConstantPacer
```

Comment:
OK, fair enough. I went with this approach because I expected the type alias to have a limited lifetime, but if you're wanting to keep it around in perpetuity then using the Rate alias here does make sense.